src/app/component/CategoryProductList/CategoryProductList.component.js

```diff
@@ -22,154 +24,236 @@ import './CategoryProductList.style';
+
+        if (isLoading) {
+            this.pagesIntersecting = [];
+        }
+
+        if (!this.observer && 'IntersectionObserver' in window) {
+            const options = {
+                rootMargin: '0px',
+                threshold: 0.1
+            };
+
+            this.observer = new IntersectionObserver((entries) => {
+                const pageFromUrl = parseInt(getQueryParam('page', location) || 1, 10);
+
+                entries.forEach((entry) => {
+                    const page = parseInt(
+                        Object.keys(this.nodes).find(key => this.nodes[key] === entry.target),
+                        10
+                    );
+
+                    const index = this.pagesIntersecting.indexOf(page);
+
+                    if (entry.isIntersecting) {
+                        if (index === -1) this.pagesIntersecting.push(page);
+                    } else if (index > -1) {
+                        this.pagesIntersecting.splice(index, 1);
+                    }
+                });
+
+                const minPage = Math.min(...this.pagesIntersecting);
+                if (minPage < Infinity && minPage !== pageFromUrl) {
+                    updatePage(minPage);
+                }
+            }, options);
         }
+        this.updateObserver();
     }
 
-    stopObserving() {
+    componentWillUnmount() {
         if (this.observer) {
-            if (this.observer.unobserve) {
-                this.observer.unobserve(this.node);
-            }
             if (this.observer.disconnect) {
                 this.observer.disconnect();
             }
+
             this.observer = null;
         }
     }
 
     /**
-     * Increase page count and update previous items length
+     * Get boundaries for pages.
+     * @return {{maxPage:Number, minPage:Number, loadedPagesCount:Number}}
      */
-    showLoading() {
-        const { items, increasePage } = this.props;
-        this.setState({ prevItemsLength: items.length });
-        increasePage();
-    }
+    getPagesBounds() {
+        const { pages } = this.props;
 
-    renderNoProducts() {
-        return (
-            <div
-              block="CategoryProductList"
-              elem="NoProducts"
-            >
-                No products found
-            </div>
-        );
+        const keys = Object.keys(pages);
+
+        const maxPage = Math.max(...keys);
+        const minPage = Math.min(...keys);
+
+        const loadedPagesCount = keys.length;
+
+        return { maxPage, minPage, loadedPagesCount };
     }
 
-    renderProducts() {
-        const { items, customFilters } = this.props;
+    /**
+     * Observes new pages and unobserves obsolete
+     */
+    updateObserver() {
+        const currentNodes = Object.values(this.nodes);
+
+        if (!this.observer || currentNodes.length <= 0) return;
 
-        if (items.length === 0) return this.renderNoProducts();
+        currentNodes.forEach((node) => {
```

Comment:
Might it be possible to reduce currentNodes to a this.observedNodes imidiatelly, without looping through again on line 139?

Reply:
Unfortunately - no, while `currentNodes` does contain new nodes (from infiniteLoader or new page), observed Nodes has no clue about them and vice-versa (like currentNodes will not cover the ones that should be unobserved). The solution may look differently (ie with double `reduce` usage) but it will still require 2 loops.